rust-symcrypt/src/hash.rs

```diff
@@ -1,59 +1,98 @@
+/* Hashing functions. For further documentation please refer to symcrypt.h */
+
 use core::ffi::c_void;
 use std::mem;
+use std::pin::Pin;
 use std::ptr;
 use symcrypt_sys;
 
 pub const SHA256_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA256_RESULT_SIZE as usize;
 pub const SHA384_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA384_RESULT_SIZE as usize;
 
-pub trait Hash {
+/// Generic trait for stateful hashing
+///
+/// append() appends to be hashed data to the state, this operation can be done multiple times.
+///
+/// result() returns the result of the hash. Once result() is called, the lifetime of the ShaXXXState is finished.
+/// To perform other stateful hash operation you must create a new hash object via ShaXXXState::new()
+///
+/// copy() creates a copy of the current ShaXXXState
+pub trait HashState {
     type Result;
 
     fn append(&mut self, data: &[u8]);
-    fn result(&mut self) -> Self::Result;
+    fn result(self) -> Self::Result;
+    fn copy(&self) -> Box<Self>;
```

Comment:
From a Rust standpoint, the `copy` function strikes me as a little weird. In Rust a copy is usually a bitwise copy (if your type impls the `Copy` trait that means the compiler can safely duplicate it with `memcpy`), and `clone()` is used for more complex copies. That said, I wouldn't just rename `copy` to `clone`, since `clone` is expected to come from the `Clone` trait instead.

One option is to redeclare `HashState` as `pub trait HashState: Clone`, which means everything that implements `HashState` can also be cloned. The problem is the way you've written this now the copy is always in a box and we wouldn't be able to guarantee that with `pub trait HashState: Clone`.

Another option is to have the `HashXXXState`s all `impl Clone for Box<HashXXXState>`, which would let you clone a box of a hash state and guarantee the copy is in a box.

Anyway, it's probably better for now to leave things mostly like you have them. Maybe rename it to `duplicate` or something to avoid the semantic clash with `copy`. I'll leave this up to your judgment though.

Reply:
After seeing how `copy` is implemented and how the various `ShaXXXState` structs work, I think I would suggest just using `Clone` instead. You don't need to keep `Self` in a box here, because your structs all add in internal indirection to keep the pinning guarantees that are needed. This means the clone impl can create any new memory allocations and such that are needed. Let me know if that makes sense!

Reply:
I see that rustls used fork for this function name.

Reply:
> After seeing how `copy` is implemented and how the various `ShaXXXState` structs work, I think I would suggest just using `Clone` instead. You don't need to keep `Self` in a box here, because your structs all add in internal indirection to keep the pinning guarantees that are needed. This means the clone impl can create any new memory allocations and such that are needed. Let me know if that makes sense!

This makes sense, I have moved the copy over to.

`impl Clone for ShaXXXState` 

Unit tests work so I think it should be fine.